plugin/firewall/setup.go

```diff
@@ -0,0 +1,178 @@
+		}
+		// check if location already used or not
+		for c.NextBlock() {
+			r, err := p.parseOptionOrRule(c)
+			if err != nil {
+				return nil, err
+			}
+			rl.addRuleElement(r)
+		}
+	}
+	return p, nil
+}
+
+func (p *firewallPlugin) parseOptionOrRule(c *caddy.Controller) (*ruleElement, error) {
+	// by default, at least one engine is available : the ExpressionEngine
+	e := &policy.ExpressionEngine{}
+	switch c.Val() {
+	// check each possible option of the plugin
+
+	case policy.NameTypes[policy.TypeRefuse]:
+		fallthrough
+	case policy.NameTypes[policy.TypeAllow]:
+		fallthrough
+	case policy.NameTypes[policy.TypeBlock]:
+		fallthrough
+	case policy.NameTypes[policy.TypeDrop]:
+		// these 4 direct policy action denotes the default ExpressionEngine
+		action := c.Val()
+		name := "--default--"
+		args := c.RemainingArgs()
+		if len(args) < 1 {
+			return nil, fmt.Errorf("not enough arguments to build a policy rule, expect allow/refuse/block/drop query/reply <expression>, got %s %s", c.Val(), strings.Join(args, " "))
+		}
+		params := append([]string{action}, args...)
+		r, err := e.BuildRule(params)
+		if err != nil {
+			return nil, err
+		}
+		return &ruleElement{"", name, params, r}, nil
+
+	default:
+		// we can only suppose it is an engine type(plugin name), name and args
+		plugin := c.Val()
+		args := c.RemainingArgs()
+		if len(args) < 1 {
+			return nil, fmt.Errorf("not enough arguments to build a policy rule, expect %s <name-engine>", c.Val())
+		}
+		name := args[0]
+		params := args[1:]
+		// as the Engine are not yet knowm, just create a ruleElement with the parameters.The Rule will be created later
+		return &ruleElement{plugin, name, params, nil}, nil
+
+	}
+}
+
+func (p *firewallPlugin) enrollEngines(c *caddy.Controller) error {
+
+	var eng = make(map[string]map[string]string)
+	// build a Map of missing Engines needed to build all rules of the RuleLists
+	for _, loc := range []*ruleList{p.query, p.reply} {
+		for _, re := range loc.ruleList {
+			if _, ok := p.engines[re.name]; !ok {
+				names, ok := eng[re.plugin]
+				if !ok {
+					names = make(map[string]string)
+					eng[re.plugin] = names
+				}
+				if _, ok := names[re.name]; !ok {
+					names[re.name] = re.name
+				}
+			}
+		}
+	}
+
+	// retrieve all needed Engine.
+	// These are plugins that implements the 'Engineer' interface
+	plugins := dnsserver.GetConfig(c).Handlers()
+	for _, pl := range plugins {
+		if e, ok := pl.(Engineer); ok {
+			if names, okn := eng[pl.Name()]; okn {
+				for n := range names {
+					re := e.GetEngine(n)
+					if re == nil {
+						return c.Errf("missing policy engine for plugin %s and name %s", p.Name(), n)
+					}
+					p.engines[n] = re
+					delete(names, n)
+				}
+			}
+		}
+	}
+
+	// build a list of all engines not found
+	allnames := make([]string, 0)
+	for pln, names := range eng {
+		for n := range names {
+			allnames = append(allnames, pln+"/"+n)
```

Comment:
you can break here and report the erorr, removing the need for `allnames` and friends alltogther.

Reply:
Ok, but we will report only one error at each setup ..